File: token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
}

const (
	Illegal = "Illegal" // unknown token
	Eof     = "Eof"     // end of file

	// identifiers and literals
	Ident = "Ident" // named var/fns
	Int   = "Int"
  String = "String"

	// Ops
	Assign   = "="
	Plus     = "+"
	Minus    = "-"
	Bang     = "!"
	Asterisk = "*"
	Slash    = "/"
	Lt       = "<"
	Gt       = ">"
	Eq       = "=="
	NEq      = "!="

	// Delims
	Comma     = "," // var delimiter
	Semicolon = ";" // line end (along with \n)

	// Scopes
	LParen = "("
	RParen = ")"
	LBrace = "{"
	RBrace = "}"

	// Keywords
	Function = "Function"
	Let      = "Let"
	If       = "If"
	Else     = "Else"
	True     = "True"
	False    = "False"
	Return   = "Return"
)

var keywords = map[string]TokenType{
	"fn":     Function,
	"let":    Let,
	"if":     If,
	"else":   Else,
	"true":   True,
	"false":  False,
	"return": Return,
}

func LookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return Ident
}

File: rust_src/ast/src/token.rs
use std::collections::HashMap;
use crate::TokenKind::{Function, Let};

#[derive(Debug, Eq, PartialEq)]
pub enum TokenKind {
    // terminating tokens
    Illegal,
    Eof,

    // identifiers / literals
    Identifier,
    Integer,
    True,
    False,

    // operators
    Bind,
    Plus,
    Minus,
    Bang,
    Aster,
    Slash,
    Lt,
    Gt,
    Eq,
    NEq,

    // delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,

    // keywords
    Function,
    Let,
    If,
    Else,
    Return,
}

#[derive(Debug)]
pub struct Token {
    pub kind: TokenKind,
    pub literal: Box<str>,
}

impl Default for Token {
    fn default() -> Token {
        Token {
            kind: TokenKind::Illegal,
            literal: Box::from(""),
        }
    }
}

impl Token {
    pub fn lookup_identifier<S: AsRef<str>>(id: S) -> TokenKind {
        Self::get_keyword(id).unwrap_or(TokenKind::Identifier)
    }

    pub fn get_keyword<S: AsRef<str>>(kw: S) -> Option<TokenKind> {
        match kw.as_ref() {
            "fn" => Some(TokenKind::Function),
            "let" => Some(TokenKind::Let),
            "true" => Some(TokenKind::True),
            "false" => Some(TokenKind::False),
            "return" => Some(TokenKind::Return),
            "if" => Some(TokenKind::If),
            "else" => Some(TokenKind::Else),
            _ => None,
        }
    }
}